git/gitattr/tree_test.go

```diff
@@ -0,0 +1,246 @@
+					Attrs: []*Attr{
+						{
+							K: "subdir", V: "yes",
+						},
+					},
+				}},
+			},
+		},
+	}
+)
+
+func TestTreeAppliedInRoot(t *testing.T) {
+	attrs := example.Applied("a.dat")
+
+	assert.Len(t, attrs, 4)
+	assert.Equal(t, attrs[0], &Attr{K: "filter", V: "lfs"})
+	assert.Equal(t, attrs[1], &Attr{K: "diff", V: "lfs"})
+	assert.Equal(t, attrs[2], &Attr{K: "merge", V: "lfs"})
+	assert.Equal(t, attrs[3], &Attr{K: "text", V: "false"})
+}
+
+func TestTreeAppliedInSubtreeRelevant(t *testing.T) {
+	attrs := example.Applied("subdir/a.dat")
+
+	assert.Len(t, attrs, 5)
+	assert.Equal(t, attrs[0], &Attr{K: "filter", V: "lfs"})
+	assert.Equal(t, attrs[1], &Attr{K: "diff", V: "lfs"})
+	assert.Equal(t, attrs[2], &Attr{K: "merge", V: "lfs"})
+	assert.Equal(t, attrs[3], &Attr{K: "text", V: "false"})
+	assert.Equal(t, attrs[4], &Attr{K: "subdir", V: "yes"})
+}
+
+func TestTreeAppliedInSubtreeIrrelevant(t *testing.T) {
+	attrs := example.Applied("subdir/a.txt")
+
+	assert.Empty(t, attrs)
+}
+
+func TestTreeAppliedInIrrelevantSubtree(t *testing.T) {
+	attrs := example.Applied("other/subdir/a.dat")
+
+	assert.Len(t, attrs, 4)
+	assert.Equal(t, attrs[0], &Attr{K: "filter", V: "lfs"})
+	assert.Equal(t, attrs[1], &Attr{K: "diff", V: "lfs"})
+	assert.Equal(t, attrs[2], &Attr{K: "merge", V: "lfs"})
+	assert.Equal(t, attrs[3], &Attr{K: "text", V: "false"})
+}
+
+func TestNewDiscoversSimpleTrees(t *testing.T) {
+	tmp, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer os.Remove(tmp)
+
+	db, err := gitobj.FromFilesystem(tmp, "")
+	require.NoError(t, err)
+	defer db.Close()
+
+	blob, err := db.WriteBlob(gitobj.NewBlobFromBytes([]byte(`
+		*.dat filter=lfs diff=lfs merge=lfs -text
+	`)))
+	require.NoError(t, err)
+
+	tree, err := New(db, &gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     ".gitattributes",
+			Oid:      blob,
+			Filemode: 0100644,
+		},
+	}})
+	require.NoError(t, err)
+
+	attrs := tree.Applied("foo.dat")
+
+	assert.Len(t, attrs, 4)
+	assert.Equal(t, attrs[0], &Attr{K: "filter", V: "lfs"})
+	assert.Equal(t, attrs[1], &Attr{K: "diff", V: "lfs"})
+	assert.Equal(t, attrs[2], &Attr{K: "merge", V: "lfs"})
+	assert.Equal(t, attrs[3], &Attr{K: "text", V: "false"})
+}
+
+func TestNewDiscoversSimpleChildrenTrees(t *testing.T) {
+	tmp, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer os.Remove(tmp)
+
+	db, err := gitobj.FromFilesystem(tmp, "")
+	require.NoError(t, err)
+	defer db.Close()
+
+	blob, err := db.WriteBlob(gitobj.NewBlobFromBytes([]byte(`
+		*.dat filter=lfs diff=lfs merge=lfs -text
+	`)))
+	require.NoError(t, err)
+
+	child, err := db.WriteTree(&gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     ".gitattributes",
+			Oid:      blob,
+			Filemode: 0100644,
+		},
+	}})
+
+	tree, err := New(db, &gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     "child",
+			Oid:      child,
+			Filemode: 040000,
+		},
+	}})
+	require.NoError(t, err)
+
+	attrs := tree.Applied("child/foo.dat")
```

Comment:
Since `child/foo.dat` would be matched by a `*.dat` pattern at the top level, it might be worth explicitly asserting here that that `tree` has no attributes, but one child with 4 attributes

Reply:
Good call! Additionally, both `Lines` and `Children` are part of the public state, so we should check that they match what we expect them to. I did that in https://github.com/git-lfs/git-lfs/pull/3114/commits/ff62456c555e0396b24585ec907fd56784469e1d.